ratatool-sampling/src/main/scala/com/spotify/scio/Random.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2018 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio
+
+import com.spotify.scio.util.random.RandomSampler
+
+import java.util.concurrent.ConcurrentHashMap
+
+import org.apache.beam.sdk.transforms.DoFn
+import org.apache.beam.sdk.transforms.DoFn.ProcessElement
+import java.util.{Random => JRandom}
+
+object Random {
+
+  /**
+   * Modified version of Scio RandomValueSampler to allow sampling on a per-element probability
+   */
+  abstract class RandomValueSampler[K, V, R]()
+    extends DoFn[((K, V), Double), (K, V)] {
+
+    protected lazy val rngs: ConcurrentHashMap[K, R] = new ConcurrentHashMap[K, R]()
+    protected var seed: Long = -1
+
+    @ProcessElement
+    def processElement(c: DoFn[((K, V), Double), (K, V)]#ProcessContext): Unit = {
+      val ((key, value), fraction) = c.element()
+
+      rngs.putIfAbsent(key, init(fraction))
+      val count = samples(fraction, rngs.get(key))
+      var i = 0
+      while (i < count) {
+        c.output((key, value))
+        i += 1
+      }
+    }
+
+    def init(fraction: Double): R
+    def samples(fraction: Double, rng: R): Int
+    def setSeed(seed: Long): Unit = this.seed = seed
+
+  }
+
+  class BernoulliValueSampler[K, V]
```

Comment:
This currently is only a slight modification of Scio's BernoulliValueSampler where it returns the sample value. Can possibly be consolidated into Scio at a later point